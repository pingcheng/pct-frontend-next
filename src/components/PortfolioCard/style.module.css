@reference "../../app/globals.css";

.portfolioCard {
  @apply shadow-sm cursor-pointer relative rounded;
  width: 300px;
  height: 400px;
  transition: all 0.3s cubic-bezier(0.4, 0, 0.2, 1);
  transform: translateY(0) scale(1);

  &::after {
    @apply block absolute bottom-0 w-full;
    content: " ";
    height: 100px;
    background: linear-gradient(transparent, rgba(0, 0, 0, 0.5));
    border-bottom-right-radius: 5px;
    border-bottom-left-radius: 5px;
    z-index: 1;
    transition: background 0.3s ease;
  }

  &:hover {
    @apply shadow-2xl;
    transform: scale(1.03) translateY(-10px);
    
    &::after {
      background: linear-gradient(transparent, rgba(0, 0, 0, 0.7));
    }
  }
}

.imageFallback {
  @apply w-full h-full bg-gray-200 dark:bg-gray-700 flex items-center justify-center text-gray-500 dark:text-gray-400 rounded;
}
