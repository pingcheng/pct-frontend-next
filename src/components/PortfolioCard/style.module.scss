.portfolioCard {
  @apply shadow cursor-pointer transition ease-out relative rounded;
  width: 300px;
  height: 400px;

  &::after {
    @apply block absolute bottom-0 w-full;
    content: " ";
    height: 100px;
    background: linear-gradient(transparent,rgba(0,0,0,.5));
    border-bottom-right-radius: 5px;
    border-bottom-left-radius: 5px;
  }

  &:hover {
    @apply shadow-2xl;
    transform: scale(1.03) translateY(-10px);
  }
}