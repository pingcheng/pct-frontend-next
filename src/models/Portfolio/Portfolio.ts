export class Portfolio {
  slug: string;
  name: string;
  coverImage: string;
  url: string | null;

  shortDescription: string;
  longDescription: string;

  workplace: string;
  projectRole: string;
  roleDescription: string[];

  members: string[];
  screenshots: string[];

  constructor(slug: string, props: PortfolioProps) {
    this.slug = slug;
    this.name = props.name ?? "";
    this.coverImage = props.coverImage ?? "";
    this.url = props.url ?? null;

    this.shortDescription = props.shortDescription ?? "";
    this.longDescription = props.longDescription ?? "";

    this.workplace = props.workplace ?? "";
    this.projectRole = props.projectRole ?? "";
    this.roleDescription = props.roleDescription ?? [];

    this.members = props.members ?? [];
    this.screenshots = props.screenshots ?? [];
  }

  get hasScreenshots(): boolean {
    return this.screenshots.length > 0;
  }
}

export type PortfolioProps = {
  name: string | null;
  coverImage: string | null;
  url: string | null;

  shortDescription: string | null;
  longDescription: string | null;

  workplace: string | null;
  projectRole: string | null;
  roleDescription: string[] | null;

  members: string[] | null;
  screenshots: string[] | null;
};
