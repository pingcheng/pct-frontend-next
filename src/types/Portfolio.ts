export type { Portfolio } from "@/models/Portfolio/Portfolio";
export type { Portfolio as SerializedPortfolio } from "@/models/Portfolio/Portfolio";