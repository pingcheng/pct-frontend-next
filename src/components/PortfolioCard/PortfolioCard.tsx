"use client";

import Image from "next/image";
import { useState, useRef } from "react";
import { useRouter } from "next/navigation";
import { Portfolio } from "@/models/Portfolio/Portfolio";
import styles from "./style.module.css";

type PortfolioCardProps = {
  portfolio: Portfolio;
  priority?: boolean;
};

export default function PortfolioCard({
  portfolio,
  priority = false,
}: PortfolioCardProps) {
  const [imageError, setImageError] = useState(false);
  const [transform, setTransform] = useState("rotateX(0deg) rotateY(0deg)");
  const containerRef = useRef<HTMLDivElement>(null);
  const router = useRouter();

  const handleMouseMove = (e: React.MouseEvent<HTMLDivElement>) => {
    if (!containerRef.current) return;
    
    const rect = containerRef.current.getBoundingClientRect();
    const x = e.clientX - rect.left;
    const y = e.clientY - rect.top;
    
    const centerX = rect.width / 2;
    const centerY = rect.height / 2;
    
    const rotateX = (y - centerY) / centerY * -10;
    const rotateY = (x - centerX) / centerX * 10;
    
    setTransform(`rotateX(${rotateX}deg) rotateY(${rotateY}deg)`);
  };

  const handleMouseLeave = () => {
    setTransform("rotateX(0deg) rotateY(0deg)");
  };

  const handleClick = () => {
    router.push(`/portfolio/${portfolio.slug}`);
  };

  if (imageError) {
    return (
      <div
        className={styles.imageFallback}
        role="img"
        aria-label={`Cover image for ${portfolio.name} project (image unavailable)`}
      >
        <div
          className="absolute bottom-0 p-2 text-gray-200"
          style={{ zIndex: 10 }}
        >
          <div className="text-sm font-bold">{portfolio.name}</div>
          <div className="text-xs">{portfolio.shortDescription}</div>
        </div>
      </div>
    );
  }

  return (
    <div 
      ref={containerRef}
      className={styles.perspectiveContainer}
      onMouseMove={handleMouseMove}
      onMouseLeave={handleMouseLeave}
      onClick={handleClick}
    >
      <div
        className={styles.portfolioCard}
        style={{
          backgroundImage: `url('${portfolio.coverImage}')`,
          backgroundSize: "cover",
          transform: transform,
        }}
        role="article"
        aria-label={`${portfolio.name} project`}
        onError={() => setImageError(true)}
      >
        <div
          className="absolute bottom-0 p-2 text-gray-200"
          style={{ zIndex: 10 }}
        >
          <div className="text-sm font-bold">{portfolio.name}</div>
          <div className="text-xs">{portfolio.shortDescription}</div>
        </div>
      </div>
    </div>
  );
}
