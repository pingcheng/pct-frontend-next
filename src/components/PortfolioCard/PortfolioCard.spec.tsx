import { render } from "@testing-library/react";
import PortfolioCard from "@/components/PortfolioCard/PortfolioCard";
import { Portfolio } from "@/models/Portfolio/Portfolio";

describe("test PortfolioCard", () => {
  const portfolio: Portfolio = {
    slug: "test-portfolio",
    name: "Test Portfolio",
    coverImage: "https://sample.jpg",
    url: "https://sample.com",
    shortDescription: "This is test portfolio description",
    longDescription: "This is the long version of portfolio description",
    workplace: "Ping Cheng Dev Space",
    projectRole: "Developer",
    roleDescription: ["Design", "Development", "Test"],
    members: ["Ping Cheng"],
    screenshots: ["https://image1.jpg", "https://image2.jpg"],
    hasScreenshots: true,
  };

  it("should match the snapshot", () => {
    const { container } = render(<PortfolioCard portfolio={portfolio} />);
    expect(container).toMatchSnapshot();
  });
});
