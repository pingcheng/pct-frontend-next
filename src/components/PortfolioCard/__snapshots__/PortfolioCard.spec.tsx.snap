// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`test PortfolioCard should handle priority prop 1`] = `
<div>
  <div
    class="perspectiveContainer"
  >
    <div
      aria-label="Test Portfolio project"
      class="portfolioCard"
      role="article"
      style="background-image: url("https://sample.jpg"); background-size: cover; transform: rotateX(0deg) rotateY(0deg);"
    >
      <div
        class="absolute bottom-0 p-2 text-gray-200"
        style="z-index: 10;"
      >
        <div
          class="text-sm font-bold"
        >
          Test Portfolio
        </div>
        <div
          class="text-xs"
        >
          This is test portfolio description
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test PortfolioCard should match the snapshot 1`] = `
<div>
  <div
    class="perspectiveContainer"
  >
    <div
      aria-label="Test Portfolio project"
      class="portfolioCard"
      role="article"
      style="background-image: url("https://sample.jpg"); background-size: cover; transform: rotateX(0deg) rotateY(0deg);"
    >
      <div
        class="absolute bottom-0 p-2 text-gray-200"
        style="z-index: 10;"
      >
        <div
          class="text-sm font-bold"
        >
          Test Portfolio
        </div>
        <div
          class="text-xs"
        >
          This is test portfolio description
        </div>
      </div>
    </div>
  </div>
</div>
`;
